image/manifest.go

```diff
@@ -26,28 +26,274 @@ import (
+// given parameters. The manifest will only allow one sequence per validation set,
+// meaning that any subsequent calls to this will be ignored.
+func (sm *SeedManifest) SetAllowedValidationSet(accountID, name string, sequence int, pinned bool) error {
+	if sequence <= 0 {
+		return fmt.Errorf("cannot add allowed validation set %q for a unknown sequence",
+			fmt.Sprintf("%s/%s", accountID, name))
+	}
+
+	vs := &SeedManifestValidationSet{
+		AccountID: accountID,
+		Name:      name,
+		Sequence:  sequence,
+		Pinned:    pinned,
+	}
+	if _, ok := sm.vsAllowed[vs.Unique()]; !ok {
+		sm.vsAllowed[vs.Unique()] = vs
+	}
+	return nil
+}
+
+// MarkSnapRevisionSeeded attempts to mark a snap-revision as seeded in the manifest.
+// The seeded revision will be validated against any previously allowed revisions set. It
+// will also be validated against any revisions set in previously seeded validation sets.
+func (sm *SeedManifest) MarkSnapRevisionSeeded(snapName string, revision int) error {
+	snapRev := snap.R(revision)
+	if rev, ok := sm.revsAllowed[snapName]; ok {
+		// Allowed revision specified, it must match.
+		if rev.Revision != snapRev {
+			return fmt.Errorf("revision %s does not match the allowed revision %s", snapRev, rev.Revision)
+		}
+	}
+
+	if rev, ok := sm.revsSeeded[snapName]; ok {
+		// Already marked as seeding.
+		return fmt.Errorf("cannot mark %q (%s) as seeded, it has already been marked seeded: %q (%s)",
+			snapName, snapRev, rev.SnapName, rev.Revision)
+	}
+
+	sm.revsSeeded[snapName] = &SeedManifestSnapRevision{
+		SnapName: snapName,
+		Revision: snapRev,
+	}
+	return nil
+}
+
+// MarkValidationSetSeeded marks a validation-set as seeded. It verifies against any previously
+// set rules by SetAllowedValidationSet, and sets up new rules based on the snaps defined in the
+// validation set.
+// This relies on validation-set assertions being marked here the moment they are fetched by
+// the seedwriter, which should be done before the first call to MarkSnapRevisionSeeded.
+func (sm *SeedManifest) MarkValidationSetSeeded(vsa *asserts.ValidationSet, pinned bool) error {
+	vs := newSeedManifestValidationSet(vsa, pinned)
+	if _, ok := sm.vsSeeded[vs.Unique()]; ok {
+		return fmt.Errorf("cannot mark validation set %q as seeded, it has already been marked as such",
+			vs.Unique())
+	}
+
+	// Check against any pre-defined restrictions.
+	if allowed, ok := sm.vsAllowed[vs.Unique()]; ok {
+		if allowed.Sequence != vs.Sequence {
+			return fmt.Errorf("sequence of %q (%d) does not match the allowed sequence (%d)",
+				vs.Unique(), vs.Sequence, allowed.Sequence)
+		}
+		if allowed.Pinned != pinned {
+			return fmt.Errorf("pinning of %q (%t) does not match the allowed pinning (%t)",
+				vs.Unique(), pinned, allowed.Pinned)
+		}
+	}
+
+	for _, sn := range vsa.Snaps() {
+		if sn.Presence == asserts.PresenceInvalid {
+			continue
+		}
+
+		// Update allowed snaps based on the validation-set.
```

Comment:
sn.Revision could be 0 meaning not fixed which probably means we shouldn't record that and anyway atm that would fail, so we are missing a test about that

Reply:
Ah right! Thank you for catching that, of course having snaps specified in validation sets does not mean it always has a revision. And thank you, added the missing unit test for both missing revision and having 'invalid' presence